avm/res/network/application-gateway/main.bicep

```diff
@@ -0,0 +1,608 @@
+      trustedRootCertificates: trustedRootCertificates
+      urlPathMaps: urlPathMaps
+    },
+    (enableFips
+      ? {
+          enableFips: enableFips
+        }
+      : {}),
+    (!empty(webApplicationFirewallConfiguration)
+      ? { webApplicationFirewallConfiguration: webApplicationFirewallConfiguration }
+      : {})
+  )
+  zones: zones
+}
+
+resource applicationGateway_lock 'Microsoft.Authorization/locks@2020-05-01' =
+  if (!empty(lock ?? {}) && lock.?kind != 'None') {
+    name: lock.?name ?? 'lock-${name}'
+    properties: {
+      level: lock.?kind ?? ''
+      notes: lock.?kind == 'CanNotDelete'
+        ? 'Cannot delete resource or child resources.'
+        : 'Cannot delete or modify the resource or child resources.'
+    }
+    scope: applicationGateway
+  }
+
+resource applicationGateway_diagnosticSettings 'Microsoft.Insights/diagnosticSettings@2021-05-01-preview' = [
+  for (diagnosticSetting, index) in (diagnosticSettings ?? []): {
+    name: diagnosticSetting.?name ?? '${name}-diagnosticSettings'
+    properties: {
+      storageAccountId: diagnosticSetting.?storageAccountResourceId
+      workspaceId: diagnosticSetting.?workspaceResourceId
+      eventHubAuthorizationRuleId: diagnosticSetting.?eventHubAuthorizationRuleResourceId
+      eventHubName: diagnosticSetting.?eventHubName
+      metrics: diagnosticSetting.?metricCategories ?? [
+        {
+          category: 'AllMetrics'
+          timeGrain: null
+          enabled: true
+        }
+      ]
+      logs: diagnosticSetting.?logCategoriesAndGroups ?? [
+        {
+          categoryGroup: 'AllLogs'
+          enabled: true
+        }
+      ]
+      marketplacePartnerId: diagnosticSetting.?marketplacePartnerResourceId
+      logAnalyticsDestinationType: diagnosticSetting.?logAnalyticsDestinationType
+    }
+    scope: applicationGateway
+  }
+]
+
+module applicationGateway_privateEndpoints '../../network/private-endpoint/main.bicep' = [
+  for (privateEndpoint, index) in (privateEndpoints ?? []): {
+    name: '${uniqueString(deployment().name, location)}-applicationGateway-PrivateEndpoint-${index}'
+    scope: resourceGroup(privateEndpoint.?resourceGroupName ?? '')
+    params: {
+      // Variant 1: A default service can be assumed (i.e., for services that only have one private endpoint type)
+      name: privateEndpoint.?name ?? 'pep-${last(split(applicationGateway.id, '/'))}-${privateEndpoint.?service ?? '>defaultServiceName<'}-${index}'
+      privateLinkServiceConnections: privateEndpoint.?isManualConnection != true
+        ? [
+            {
+              name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(applicationGateway.id, '/'))}-${privateEndpoint.?service ?? '>defaultServiceName<'}-${index}'
+              properties: {
+                privateLinkServiceId: applicationGateway.id
+                groupIds: [
+                  privateEndpoint.?service ?? '>defaultServiceName<'
```

Comment:
Beware, '>defaultServiceName<' needs updating and at all other locations used.

Also, note this variant of the schema should only be used if the value of `service` can only be one (like `vault` for `Key Vault`). If it can take more values, you shuold use the 2nd schema defined in the specs, that essentially does not have a default and also enforces the property as required in the UDT.

For this one it 'may' be only one allowed value as per the following reference, but please doublecheck: https://blog.blksthl.com/2023/03/22/the-complete-list-of-groupids-for-private-endpoint-privatelink-service-connection/

Reply:
@AlexanderSehr That blog links to this doc: https://github.com/MicrosoftDocs/azure-docs/blob/main/articles/private-link/private-endpoint-overview.md#private-link-resource. This doc says that the sub-resource for application gateway is "Frontend IP Configuration name". The module does not require a frontend IP configuration by default. However, it does only support one groupID AFAICT. 

Not sure what to set for `>defaultServiceName<` though. Do you think something like `appgw`?

Reply:
Hey @ilhaan I just deployed the service to see what services I could select but it doesn't work the same way as other private endpoint, asking to select specific sub-resources that I did not have available (and I don't know which they are).

Given that we do deploy a private endpoint in e.g. the waf-aligned test and it's value is `public` (which I would not have guessed), I'd take a guess and say that presumably there are several different values we can set.

The good news is that immediately renders the schema that you used invalid and we should use the one that requires the user to provide a value. You can find it as `Variant 2` [here](https://azure.github.io/Azure-Verified-Modules/specs/shared/interfaces/#private-endpoints).

Reply:
@AlexanderSehr Updated: https://github.com/Azure/bicep-registry-modules/pull/835/commits/51cec28f72e0b51858f01e4ae11f46ff76c65487. And pipeline is green.